lib/wagn/migration.rb

```diff
@@ -1,34 +1,59 @@
 # -*- encoding : utf-8 -*-
 
 class Wagn::Migration < ActiveRecord::Migration
-  def self.find_unused_name base_name
-    test_name = base_name
-    add = 1
-    while Card.exists?(test_name) do
-      test_name = "#{base_name}#{add}"
-      add +=1
+  @type = :deck_cards
+
+  class << self
+    
+    # Rake tasks use class methods, migrations use instance methods.
+    # To avoid repetition a lot of instance methods here just call class methods.
+    # The subclass Wagn::CoreMigration needs a different @type so we can't use a
+    # class variable @@type. It has to be a class instance variable.
+    # Migrations are subclasses of Wagn::Migration or Wagn::CoreMigration but they
+    # don't inherit the @type. The method below solves this problem.
+    def type 
+      @type || (ancestors[1] && ancestors[1].type)
+    end
+    
+    def find_unused_name base_name
+      test_name = base_name
+      add = 1
+      while Card.exists?(test_name) do
+        test_name = "#{base_name}#{add}"
+        add +=1
+      end
+      test_name
     end
-    test_name
-  end
 
-  def self.paths type
-    Wagn.paths["db/migrate#{schema_suffix type}"].to_a
-  end
+    def paths mig_type=type
+      Wagn.paths["db/migrate#{schema_suffix mig_type}"].to_a
+    end
   
-  def self.schema_suffix type
-    Wagn::Version.schema_suffix type
-  end
+    def schema_suffix mig_type=type
+      Wagn::Version.schema_suffix( mig_type )
+    end
   
-  def self.schema_mode type
-    new_suffix = Wagn::Migration.schema_suffix type
-    original_suffix = ActiveRecord::Base.table_name_suffix
+    def schema_mode mig_type=type
+      new_suffix = schema_suffix mig_type
+      original_suffix = ActiveRecord::Base.table_name_suffix
+    
+      ActiveRecord::Base.table_name_suffix = new_suffix
+      yield
+      ActiveRecord::Base.table_name_suffix = original_suffix
+    end
     
-    ActiveRecord::Base.table_name_suffix = new_suffix
-    yield
-    ActiveRecord::Base.table_name_suffix = original_suffix
-  end
-  
   
+    def data_path filename=nil
+      if filename
+        self.paths.each do |path|
+          path_to_file = File.join path, 'data', filename
+          return path_to_file if File.exists? path_to_file
+        end
+      else
+        File.join self.paths.first, 'data'
+      end
+    end
+  end
   
```

Comment:
I agree that it would be nice to be confident the migrations are all working after all these changes.  Any chance we can rerun the migrations from 1.10?